File: src/homemade/game/model/cellstates/SimpleState.java
package homemade.game.model.cellstates;

import homemade.game.Cell;
import homemade.game.CellState;

import java.util.EnumMap;
import java.util.EnumSet;
import java.util.Map;

public class SimpleState extends CellState {
    private static Map<Cell, CellState> simpleStates = new EnumMap<>(Cell.class);

    static {
        for (Cell type : EnumSet.of(Cell.EMPTY, Cell.MARKED_FOR_SPAWN, Cell.DEAD_BLOCK)) {
            simpleStates.put(type, new SimpleState(type));
        }
    }

    public static CellState getSimpleState(Cell type) {
        return simpleStates.get(type);
    }

    private SimpleState(Cell type) {
        super(type);
    }
}

File: game-model-v2/src/homemade/game/model/cellstates/SimpleState.kt
package homemade.game.model.cellstates

import homemade.game.model.Cell
import homemade.game.model.CellState

object SimpleState {
    private class SimpleStateImpl(type: Cell) : CellState(type)

    val allStates = setOf(Cell.EMPTY, Cell.MARKED_FOR_SPAWN, Cell.DEAD_BLOCK)
    private val simpleState = allStates.associateBy({ it }, { SimpleStateImpl(it) })

    fun get(type: Cell): CellState {
        return simpleState[type] ?: throw RuntimeException("not a simple state $type")
    }
}
